### ISSUE TITLE
Document Datagrid expandable row accessibility limitations

### ISSUE THREAD

**Issue by gnomeontherun** [vmware/clarity#3375]
The current expandable row pattern should be documented with the accessibility limitations, and our recommendation should focus on the detail pane pattern (#2005) when available. 

Here are the steps to improve the accessibility of this control.

1. Use aria-describedby to give users the following message "Screen reader table commands may not work for viewing expanded content, please use your screen reader's browse mode to read the content exposed by this button". See https://www.w3.org/TR/wai-aria-1.1/#aria-describedby for more information.
2. Use aria-controls to associate the button with the expanded row. See https://www.w3.org/TR/wai-aria-1.1/#aria-controls for more information.
3. Provide textual boundaries i.e, "beginning of expanded content" and "beginning of expanded content" for screen readers users to understand what part of the grid they are reading.

**Comment by chlane** [comment]
Here are the specifics of the how the expandable row concept breaks accessibility .

Test URL:  https://stackblitz.com/github/clane/clarityTest2/tree/master/clarity-test-app
Clarity Version: v2.0.0-rc.1
Angular Version: 8.0.0-rc.1
OS: Mac OS X 10_14_4, Windows 10
Browser: Chrome/73.0.3683.103, Safari/605.1.15, Firefox 60.6.
Screen Reader: VoiceOver, NVDA 2018.41, JAWS 2019

Description:
The clr-dg-row-detail component does not work with VoiceOver and Chrome but has partial support with other screen reader/browser combinations.

The contents of the clr-dg-row-detail component cannot be read by VoiceOver and Chrome using table commands or the reading cursor and are not read by other screen reader/browser combinations when using table navigation commands as expected.

Normally, when you read table cells with table navigation commands like Ctr + Option + arrow for VoiceOver, you hear the data along with the headers associated with that data. In the case of the details component you do not hear the data or the headers.

You can hear this content if you use the arrow key alone but it is not associated with any headers as all data within a grid is expected to be.

But screen reader users will expect that table navigation commands give them access to all the data in the table and may not realize they have to use the arrow key to find the details content.

See https://www.w3.org/TR/wai-aria-1.1/#grid

The clr-dg-row-detail component is not a row or cell which means the current implementation is invalid ARIA. Invalid ARIA can cause unexpected behaviors when using assistive technologies. Additionally, the contents of the clr-dg-row-detail component are not associated with any of the grid's headers. This makes it impossible for assistive technologies to associate cell data with the appropriate header or recognize the contents of the clr-dg-row-detail component as part of the grid.

Steps to Reproduce:
- Using VoiceOver with Chrome, go to the datagrid-expandable-rows example
- Activate one of the "datagrid-expandable-caret-button" buttons
- Try to read the contents of the newly displayed clr-dg-row-detail component using arrow keys and observe that you cannot.
- Try to read the contents of the newly displayed clr-dg-row-detail component using table navigation commands (Ctr + Option + arrow keys) and observe that you cannot.

Recommendation:
Use modality (focus management and aria-hidden or at least textual boundaries) to provide row details to screen reader users.

Additional Notes:
The details are read by NVDA 2018.41 and Firefox 60.6.are (64-bit) on Windows 10 when using the arrow keys alone. But the user will not hear header name(s) associated with any data point as expected. Additionally, when using JAWS 2019/Chrome or NVDA/Firefox, and using table specific navigation commands, i.e., CTRL + ALT + ARROW key of choice, the contents of the clr-dg-row-detail component are not read.

**Comment by chlane** [comment]
From: Sheri Byrne Haber <sbyrnehaber@vmware.com> Date: Monday, June 3, 2019 at 2:58 PM To: Chris Lane <chlane@vmware.com>, Jeremy Wilken <wilkenj@vmware.com> Cc: Jehad Affoneh <affonehj@vmware.com> Subject: RE: Instructions for expandable rows workaround
 
I think we have three options – good, better and best
 
Because we are focusing on meeting compliance minimums for now this is good enough to get us a “supports”  Better and Best can come later.
 
Sheri
 
From: Chris Lane <chlane@vmware.com>  Sent: Monday, June 3, 2019 2:57 PM To: Jeremy Wilken <wilkenj@vmware.com> Cc: Sheri Byrne Haber <sbyrnehaber@vmware.com>; Jehad Affoneh <affonehj@vmware.com> Subject: Instructions for expandable rows workaround
 
Hi Jeremy,
 
Since the workaround for screen readers is to read the expanded content with the reading cursor instead of table reading commands,  we should provide instructions to do so (great idea Sheri).
 
We can use aria-describedby on the button that triggers the expanded content to communicate this workaround to the user.
 
I think this could get us a ‘supports’ rating.  What do you think Sheri?
 
Chris

**Comment by chlane** [comment]
Here are the steps to improve the accessibility of this control.

1) Use aria-describedby to give users the following message "table commands will not work for viewing expanded content, please use your screen reader's browse mode to read the content exposed by this button".  See https://www.w3.org/TR/wai-aria-1.1/#aria-describedby for more information. 
2) Use aria-controls to bind the button with the expanded row. See  https://www.w3.org/TR/wai-aria-1.1/#aria-controls for more information. 
3) Provide textual boundaries i.e, "beginning of expanded content" and "beginning of expanded content" for screen readers users to understand what part of the grid they are reading.

**Comment by chlane** [comment]
Just pointing it out this is also a JAWS 2019/Chrome and NVDA/Firefox issue as described in the additional notes but we wont be including JAWS 2019/Chrome as one of our 'best used on' combinations so we don't have to worry about it.

**Comment by bdryanovski** [comment]
@chlane I have a question related to `aria-describedby`. 

I'm testing it on:
 * MacOS/VoiceOver - Chrome/Safari
 * Windows/NVDA - Edge/Firefox

And could not make it work. `aria-describedby` is not announced - Inside the Devtools I could see that the elements have it but VO is not reading it. From this site, it seems that the support is pretty limited when it comes to `aria-describedby` https://a11ysupport.io/tech/aria/aria-describedby_attribute.

From here https://w3c.github.io/using-aria/#label-support we could see that for MacOS VoiceOver have to be set to `role=landmark` but this also is not working.

For me is working with `role=navigation`
```html
<div role="navigation" aria-describedby="context">Expand Row area</div>
<div id="context">Additional text to be announced</div>
```
But this now is making the `Expand Row area` a navigation section, so I have additional step before I could read the content. 
 
1. Additional text to be announced, navigation
2. You are currently on navigation, inside web content,
3. Expand Row area
4. End of,  Additional text to be announced, navigation (when you try to exit the zone)

I did not test it on JAWS but it seems that it's the only one with good support for  `aria-describedby`. At least on paper.

Now the second part of my question:

`aria-controls` is not supported for VoiceOver and NVDA only JAWS. I could add the attributes but could not validate that it will work.

Did not get that: 

>Provide textual boundaries i.e, "beginning of expanded content" and "beginning of expanded content" for screen readers users to understand what part of the grid they are reading.

Did you mean  "beginning of expanded content" and "end of expanded content"

**Comment by chlane** [comment]
I’m out on PTO until Thursday let’s sync up then :)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by bdryanovski** [comment]
https://github.com/bdryanovski/clarity/pull/15 - Work in progress PR with some of the changes. I'm adding it here soo we could discuss this with code (Find it easier).

**Comment by chlane** [comment]
@bdryanovski

I just did some testing with aria-describedby and it seems to work well when using the Tab key with the browser screen reader combinations we support below.


- Internet Explorer 10 or 11 and the latest version of JAWS or NVDA
- Latest versions of Chrome, Firefox or Edge with the latest version of JAWS or NVDA
- Latest version of Safari with the latest or previous version of VoiceOver

For test results see my example.

http://cx-guidelines.vmware.com/examples/ariaDescribedBy.html

Regarding, role="navigation" I did not have to use it or any other landmark role to get aria-describedby by to work as seen in my example. But there is text that says "VO on iOS only reads it if on landmark roles" at https://w3c.github.io/using-aria/#label-support. If appropriate, consider using the region role (https://www.w3.org/TR/wai-aria-1.1/#region).

Regarding, aria-controls (https://www.w3.org/TR/wai-aria-1.1/#aria-controls), as described it is supposed to identify the element (or elements) whose contents or presence are controlled by the current element. I've never heard a screen reader make any announcements due to it's presence. Level access is claiming it is not supported by screen readers at https://labs.levelaccess.com/index.php/ARIA_Controls_Property. But I recomemend adding it, because it is used in examples from WAI-ARIA Authoring Practices 1.1 such as https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html. Adding the attribute will be enough, hopefully we will gain benefits from it in the future if screen reader vendors decide to do something with it.


Finally, about the textual boundaries I meant

"beginning of expanded content" and "end of expanded content"

as you suspected.

Thanks!

**Comment by bdryanovski** [comment]
@chlane Thanks

**Comment by clane** [comment]
2 issues were reported at http://cx-guidelines.vmware.com/clarityTesting/. Issue #10 is for the defect and issue #188 is for the documentation fo the
 defect.

Both are captured here https://github.com/vmware/clarity/issues/3375.

It is not possible to fix technically fix this issue because HTML tables can't have any elements that are not rows or cells. If they do they will not work correctly with screen reader table commands and in some cases this content cannot be read by the screen readers cursor at all. .

Since we are sticking with this pattern despite not being able to fix the defect, my recommendation was to use aria-describedby and aria-controls along with textual boundaries in case the screen reader user found the expandable row content by moving the screen reader cursor with the arrow key rrather than table commands.   

Then separately document that the expandable row content will not be accessible with some screen reader/browser combinations.

**Comment by clane** [comment]
This issue is document in section 1.3.2  of the VPAT at https://www.vmware.com/content/dam/digitalmarketing/vmware/en/pdf/product/vpat/vmware-clarity-design-system-vpat-2.2.pdf so we don't have to document it on the Clarity site. Please cancel the documentation request.
